tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,310 @@
+            // has intialized the value and closed the semaphore, in which case self.initialized
+            // is true and we don't set the value here
+            match self.semaphore.acquire().await {
+                Ok(_permit) => {
+                    if !self.initialized() {
+                        let value = f().await;
+
+                        // SAFETY: There is only one permit on the semaphore, hence only one
+                        // mutable reference is created
+                        unsafe { self.set_value(value) };
+
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!("acquired semaphore after value was already initialized");
+                    }
+                }
+                Err(_) => {
+                    if self.initialized() {
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!(
+                            "semaphore acquire call failed when value not already intialized"
+                        );
+                    }
+                }
+            }
+        }
+    }
+}
+
+// OnceCell is Send (and Sync) because access to its UnsafeCell
+// is guarded by the semaphore permit and atomic operations
+// on `value_set`, as long as T itself is Send (and Sync)
+unsafe impl<T: Sync + Send> Sync for OnceCell<T> {}
+unsafe impl<T: Send> Send for OnceCell<T> {}
+
+/// Error returned from the [`OnceCell::set`] method
+///
+/// [`OnceCell::set`]: crate::sync::OnceCell::set
+#[derive(Debug, PartialEq)]
+pub struct AlreadyInitializedError;
+
+/// Error returned from the [`OnceCell::get`] method
+///
+/// [`OnceCell::get`]: crate::sync::OnceCell::get
+#[derive(Debug, PartialEq)]
+pub struct NotInitializedError;
+
+/// Allows one to lazily call an async function
+///
+/// # Examples
+/// ```
+/// use tokio::sync::Lazy;
+/// use std::pin::Pin;
+/// use std::future::Future;
+///
+///
+/// async fn some_computation() -> u32 {
+///     1 + 1
+/// }
+///
+/// fn call_async_fn() -> Pin<Box<dyn Future<Output = u32> + Send>> {
+///     Box::pin(some_computation())
+/// }
+///
+/// #[tokio::main]
+/// async fn main() {
+///     static LAZY : Lazy<u32> = Lazy::new(call_async_fn);
+///
+///     let result = tokio::spawn(async {
+///         LAZY.get().await
+///     }).await.unwrap();
+///     assert_eq!(*result, 2);
+/// }
+/// ```
+pub struct Lazy<T, F = fn() -> Pin<Box<dyn Future<Output = T> + Send>>> {
+    cell: OnceCell<T>,
+    init: Cell<Option<F>>,
```

Comment:
Using a `Cell` in a `Sync` struct feels somewhat like we are hiding some unsafe pieces. I would prefer if this was an `UnsafeCell` so the compiler makes us actually annotate the parts of the code that do unsafe things.